compiler/src/org.graalvm.compiler.lir/src/org/graalvm/compiler/lir/RedundantInstructionElimination.java

```diff
@@ -0,0 +1,185 @@
+ * load instructions. The unused constant mov may exist after match rules.
+ */
+public final class RedundantInstructionElimination extends PreAllocationOptimizationPhase {
+
+    @Override
+    protected void run(TargetDescription target, LIRGenerationResult lirGenRes, PreAllocationOptimizationContext context) {
+        new Optimization(lirGenRes.getLIR()).apply();
+    }
+
+    private static final class Optimization {
+        private final LIR lir;
+        private final VariableMap<DefEntry> unUsedDefs;
+        private final DebugContext debug;
+
+        private Optimization(LIR lir) {
+            this.lir = lir;
+            this.debug = lir.getDebug();
+            this.unUsedDefs = new VariableMap<>();
+        }
+
+        @SuppressWarnings("try")
+        private void apply() {
+            try (Indent indent = debug.logAndIndent("RedundantInstructionElimination")) {
+                try (DebugContext.Scope s = debug.scope("CollectUnusedInstructions")) {
+                    for (AbstractBlockBase<?> b : lir.getControlFlowGraph().getBlocks()) {
+                        this.collectUnusedInBlock(b);
+                    }
+
+                    if (!unUsedDefs.isEmpty()) {
+                        debug.log("Redundant instructions that will be removed later: %s", unUsedDefs);
+                    }
+                } catch (Throwable e) {
+                    throw debug.handle(e);
+                }
+
+                try (DebugContext.Scope s = debug.scope("DeleteUnusedInstructions")) {
+                    unUsedDefs.forEach(this::deleteInstruction);
+                    // Remove unused instructions for each block.
+                    for (AbstractBlockBase<?> b : lir.getControlFlowGraph().getBlocks()) {
+                        this.removeUnusedInBlock(b);
+                    }
+                } catch (Throwable e) {
+                    throw debug.handle(e);
+                }
+            }
+        }
+
+        /**
+         * Collect unused instructions for a {@code block}.
+         */
+        private void collectUnusedInBlock(AbstractBlockBase<?> block) {
+            InstructionValueConsumer defConsumer = (instruction, value, mode, flags) -> {
+                if (isVariable(value) && isEligibleOp(instruction)) {
+                    DefEntry def = new DefEntry(block, instruction, value);
+                    unUsedDefs.put(((Variable) value), def);
+                }
+            };
+
+            InstructionValueConsumer useConsumer = (instruction, value, mode, flags) -> {
+                if (isVariable(value)) {
+                    DefEntry def = unUsedDefs.get((Variable) value);
+                    if (def != null) {
+                        assert def.value.equals(value);
+                        // Remove the instruction that defined this value from the unused map.
+                        LIRInstruction inst = def.instruction;
+                        unUsedDefs.filter(d -> (!d.instruction.equals(inst)));
```

Comment:
At some point we tried to avoid allocating lambdas. I'd replace `d -> (!d.instruction.equals(inst))` with a static method (e.g. `RedundantInstructionElimination::filterDefEntry`).

Reply:
Thanks for your review! Yeah, it makes sense. However, since `VariableMap` has the public API `public void filter(Predicate<T> predicate) ` to specially handle the filtration. I think it's better not to create another one. Another problem to use a static method in `RedundantInstructionElimination` is that the members of `VariableMap` are private and they are not expected to be public. We need to add another API to expose them if we want to go through them outside of class `VariableMap` .  But if you insist your opinion, I will try to fix it. Thanks!

Reply:
Ah, I missed that you are binding the local variable `inst`. Forget what I said.